src/esp_loader.c

```diff
@@ -261,17 +261,21 @@ esp_loader_error_t esp_loader_flash_start(uint32_t offset, uint32_t image_size,
 }
 
 
-esp_loader_error_t esp_loader_flash_write(const void *payload, uint32_t size)
+esp_loader_error_t esp_loader_flash_write(void *payload, uint32_t size)
 {
     uint32_t padding_bytes = s_flash_write_size - size;
     uint8_t *data = (uint8_t *)payload;
     uint32_t padding_index = size;
 
+    if (padding_bytes >= s_flash_write_size) {
+        return ESP_LOADER_ERROR_INVALID_PARAM;
+    }
+
     while (padding_bytes--) {
         data[padding_index++] = PADDING_PATTERN;
     }
 
-    md5_update(payload, (size + 3) & ~3);
+    md5_update(data, (size + 3) & ~3);
```

Comment:
md5_update does not modify the data, only the context, let's not cast away const-ness.

Reply:
Smile - I have argued against the fake const-ness in the function signature, but otherwise, it's OK the pass `payload` instead of `data`. I'm reverting this line.